### PR TITLE
refactor(ONNX): replaces `getValueList` helper with `createScalarSublist`

### PR DESCRIPTION
**Pull request by bjacobgordon**
A preliminary refactor to support #3945 
- extracts several new helper functions
- removes cruft

### CODE REVIEW

**Comment by bjacobgordon** [comment]
Hey, @AmosLewis, could you review this since @zjgarvey is out? Zach reviewed a lot of it under #3945, but wanted to split into two stacked PRs to make each final diff smaller.

**Comment by zjgarvey** [comment]
I'll get a chance to review today, too. But it would be good to get Chi's review too.

File: lib/Conversion/TorchOnnxToTorch/DefaultDomainQtoZ.cpp [hunk 1]
```diff
@@ -180,56 +180,67 @@ LogicalResult reduceOpImpl(OpBinder binder, ConversionPatternRewriter &rewriter,
...
+                                            selectionFromGivenVector);
+}
+
+template <typename SomeTorchScalarType>
+Value createScalarSublist(
+    /*                      at */ Location givenLoc,
+    /*                    from */ Value given1DTensor,
+    /* throughBackStartingWith */ int64_t givenIndex,
+    /*                   using */ ConversionPatternRewriter &rewriter) {
+  auto Given1DTensor = cast<Torch::BaseTensorType>(given1DTensor.getType());
```
Comment (zjgarvey) [inline lib/Conversion/TorchOnnxToTorch/DefaultDomainQtoZ.cpp:215]: Why is this called `Given1DTensor`?

**Review by zjgarvey — CHANGES_REQUESTED** [review]
Thanks, Jacob. The helper function `getValueList` would certainly benefit from some cleanup.

File: lib/Conversion/TorchOnnxToTorch/DefaultDomainQtoZ.cpp (hunk 1 above)
Comment (AmosLewis) [reply lib/Conversion/TorchOnnxToTorch/DefaultDomainQtoZ.cpp:215]: input should be enough. same for loc.

File: lib/Conversion/TorchOnnxToTorch/DefaultDomainQtoZ.cpp [hunk 2]
```diff
@@ -180,56 +180,67 @@ LogicalResult reduceOpImpl(OpBinder binder, ConversionPatternRewriter &rewriter,
...
 }
 } // namespace
 
+namespace TorchImageTensor {
+// int64_t const /*  */ batchDim = 0;
+// int64_t const /**/ channelDim = 1;
+int64_t const /* */ heightDim = 2;
+// int64_t const /*  */ widthDim = 3;
+// int64_t const /*  */ depthDim = 4;
+} // namespace TorchImageTensor
```
[the remark this answers is not in the document]
Comment (AmosLewis) [reply lib/Conversion/TorchOnnxToTorch/DefaultDomainQtoZ.cpp:242]: Delete the namespace and add some comments at `auto foremostSupportedDim = TorchImageTensor::heightDim;` to explain why foremostSupportedDim is 2 should be better.

[the remark this answers is not in the document]
Comment (AmosLewis) [reply lib/Conversion/TorchOnnxToTorch/DefaultDomainQtoZ.cpp:242]: I guess the reason add this struct is becasue the same idea used in both `onnx.Upsample` and `onnx.QuantizeLinear`, he want to globalize struct to save this value. But I think it should be fine just add 2 comments when setting it. It's not used frequently. And even you want to add it, either add in include/torch-mlir/Conversion/TorchOnnxToTorch/Utils.h files or you find some similar struct in current mlir::torch/Torch:: namespace to use.

File: lib/Conversion/TorchOnnxToTorch/DefaultDomainQtoZ.cpp (hunk 1 above)
Comment (AmosLewis) [reply lib/Conversion/TorchOnnxToTorch/DefaultDomainQtoZ.cpp:215]: Add comments and examples to explain the function before the signature will be more intuitive.

File: lib/Conversion/TorchOnnxToTorch/DefaultDomainQtoZ.cpp [hunk 3]
```diff
@@ -180,56 +180,67 @@ LogicalResult reduceOpImpl(OpBinder binder, ConversionPatternRewriter &rewriter,
...
+
+  Type SomeScalar = rewriter.getType<SomeTorchScalarType>();
+
+  Value selectionFromGivenVector = rewriter.create<Torch::AtenSelectIntOp>(
+      givenLoc, SelectionFromGivenVector, givenVector, frontDim,
+      dimOfSelection);
+
+  return rewriter.create<Torch::AtenItemOp>(givenLoc, SomeScalar,
+                                            selectionFromGivenVector);
+}
```
[the remark this answers is not in the document]
Comment (bjacobgordon) [reply lib/Conversion/TorchOnnxToTorch/DefaultDomainQtoZ.cpp:207]: Just double-checked! This is called from within `createScalarSublist`, so we're good!

File: lib/Conversion/TorchOnnxToTorch/DefaultDomainQtoZ.cpp (hunk 1 above)
Comment (bjacobgordon) [reply lib/Conversion/TorchOnnxToTorch/DefaultDomainQtoZ.cpp:215]: Great question, @zjgarvey! Since an instance of a _meta_-type is just a _dynamic_ type, there's a few options for naming:
- `{camelCaseName}Type`: communicates that it's an instance of a meta-type, but this can be ambiguous depending on the adjectives used in the name ("given", "output", "each", etc.)
- `typeOf{PascalCaseName}`: less ambiguous, but longer
- `{PascalCaseName}`: short and still clear that it's a type because of the use of PascalCase
- `{PascalCaseName}Type`: this would be a _meta_-type rather than just a type, so can't use this.

So, I named it after the source of the type using PascalCase, but I'll rename it to `Some1DTensor` to emphasize that it's more dynamic/generic and put this explanation in the commit.

Comment (zjgarvey) [reply lib/Conversion/TorchOnnxToTorch/DefaultDomainQtoZ.cpp:215]: No, I don't think pascal case is preferred here. Just use something like `tensorType` since there is only one tensor in this scope. I think there is a misunderstanding of metatypes here. The variable you are naming is an instance of a class, so just use a camel case name, preferably with the word type in it. 

In some abstract sense, `given1DTensor` is an example of a tensor which *has* the type referenced here, but the type instance you are getting is not actually a class, and `given1DTensor` is not an instance of that thing (because it's not a class).

File: lib/Conversion/TorchOnnxToTorch/DefaultDomainQtoZ.cpp [hunk 4]
```diff
@@ -180,56 +180,67 @@ LogicalResult reduceOpImpl(OpBinder binder, ConversionPatternRewriter &rewriter,
...
-        binder.getLoc(), Torch::ListType::get(Torch::IntType::get(context)),
-        itemList);
-  } else {
-    ValueList = rewriter.create<Torch::PrimListConstructOp>(
-        binder.getLoc(), Torch::ListType::get(Torch::FloatType::get(context)),
-        itemList);
+template <typename SomeTorchScalarType>
+Value createScalarAs(
+    /*    at */ Location givenLoc,
+    /*  from */ int64_t givenDim,
```
Comment (zjgarvey) [inline lib/Conversion/TorchOnnxToTorch/DefaultDomainQtoZ.cpp:186]: Might be better to call this `index` since it's purpose in this function is not as a `dim` of a tensor, but as an `index` to extract along dim 0 of a rank 1 tensor.

File: lib/Conversion/TorchOnnxToTorch/DefaultDomainQtoZ.cpp [hunk 5]
```diff
@@ -180,56 +180,67 @@ LogicalResult reduceOpImpl(OpBinder binder, ConversionPatternRewriter &rewriter,
...
-  Value ValueList;
-  if (isa<IntegerType>(xTy.getDtype())) {
-    ValueList = rewriter.create<Torch::PrimListConstructOp>(
-        binder.getLoc(), Torch::ListType::get(Torch::IntType::get(context)),
-        itemList);
-  } else {
-    ValueList = rewriter.create<Torch::PrimListConstructOp>(
-        binder.getLoc(), Torch::ListType::get(Torch::FloatType::get(context)),
-        itemList);
+template <typename SomeTorchScalarType>
```
Comment (zjgarvey) [inline lib/Conversion/TorchOnnxToTorch/DefaultDomainQtoZ.cpp:183]: This should also not be templated for similar reasons indicated in the other comment.

**Review by zjgarvey — COMMENTED** [review]
Adding a few more comments, mostly pertaining to the other function I didn't review earlier.

File: lib/Conversion/TorchOnnxToTorch/DefaultDomainQtoZ.cpp [hunk 6]
```diff
@@ -180,56 +180,67 @@ LogicalResult reduceOpImpl(OpBinder binder, ConversionPatternRewriter &rewriter,
...
+
+  Value selectionFromGivenVector = rewriter.create<Torch::AtenSelectIntOp>(
+      givenLoc, SelectionFromGivenVector, givenVector, frontDim,
+      dimOfSelection);
+
+  return rewriter.create<Torch::AtenItemOp>(givenLoc, SomeScalar,
+                                            selectionFromGivenVector);
+}
+
+template <typename SomeTorchScalarType>
```
[the remark this answers is not in the document]
Comment (bjacobgordon) [reply lib/Conversion/TorchOnnxToTorch/DefaultDomainQtoZ.cpp:209]: Excellent points, will do this while still capturing duplicated logic.

File: lib/Conversion/TorchOnnxToTorch/DefaultDomainQtoZ.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (bjacobgordon) [reply lib/Conversion/TorchOnnxToTorch/DefaultDomainQtoZ.cpp:242]: If it made you look at the namespace, @zjgarvey, then it did its job! Since ONNX and Torch differ in conventions, it's useful to state the more restrictive convention. Though I _can_ see how the wording might be confusing, so I'll tweak that.

Yeah, @AmosLewis, that's the primary reason! Comments are a great start, and usually the next level is documentative code, which usually withstands the test of time better since the compiler and test coverage get involved. Will move it to Utils.h

[the remark this answers is not in the document]
Comment (zjgarvey) [reply lib/Conversion/TorchOnnxToTorch/DefaultDomainQtoZ.cpp:242]: Please don't put this in Utils.h.

These definitions are superfluous, hyper-specific, mostly incorrect (according to <https://pytorch.org/docs/stable/generated/torch.nn.functional.interpolate.html>), and waste build time. Remove it, please.

[the remark this answers is not in the document]
Comment (bjacobgordon) [reply lib/Conversion/TorchOnnxToTorch/DefaultDomainQtoZ.cpp:242]: Okay, @zjgarvey, I'll remove it!

If we just keep it as the raw magic numbers, how do we help the newer contributors that wonder:
> Wait, why is dimension 2 the foremost resizable dim?

We _could_ comment in both places, but [comments add maintenance overhead](https://medium.com/@bpnorlander/stop-writing-code-comments-28fef5272752), that's why my first attempt was to implement self-documenting structures.

What do you think?

File: lib/Conversion/TorchOnnxToTorch/DefaultDomainQtoZ.cpp (hunk 1 above)
Comment (bjacobgordon) [reply lib/Conversion/TorchOnnxToTorch/DefaultDomainQtoZ.cpp:215]: I'll go ahead and change it, no problem! We can sync up outside of this PR to clear up any misunderstandings. 👍

File: lib/Conversion/TorchOnnxToTorch/DefaultDomainQtoZ.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (zjgarvey) [reply lib/Conversion/TorchOnnxToTorch/DefaultDomainQtoZ.cpp:242]: The word "spatial" is quite standard for non-batch/channel. Something like "firstExpectedSpatialDim" might not require too much commentary.

File: lib/Conversion/TorchOnnxToTorch/DefaultDomainQtoZ.cpp [hunk 7]
```diff
@@ -180,56 +180,67 @@ LogicalResult reduceOpImpl(OpBinder binder, ConversionPatternRewriter &rewriter,
...
-  if (isa<IntegerType>(xTy.getDtype())) {
-    ValueList = rewriter.create<Torch::PrimListConstructOp>(
-        binder.getLoc(), Torch::ListType::get(Torch::IntType::get(context)),
-        itemList);
-  } else {
-    ValueList = rewriter.create<Torch::PrimListConstructOp>(
-        binder.getLoc(), Torch::ListType::get(Torch::FloatType::get(context)),
-        itemList);
+template <typename SomeTorchScalarType>
+Value createScalarAs(
```
[the remark this answers is not in the document]
Comment (bjacobgordon) [reply lib/Conversion/TorchOnnxToTorch/DefaultDomainQtoZ.cpp:184]: Will do!

**Review by zjgarvey — COMMENTED** [review]
Nice! This looks good. Just a few changes and we should merge this. Ping me when you're done and I'll approve. 

Thanks!

File: lib/Conversion/TorchOnnxToTorch/DefaultDomainQtoZ.cpp [hunk 8]
```diff
@@ -180,53 +180,71 @@ LogicalResult reduceOpImpl(OpBinder binder, ConversionPatternRewriter &rewriter,
...
+
+  for (int indexOfEachScalar = givenIndex; indexOfEachScalar < lengthOfFullList;
+       indexOfEachScalar++) {
+    Value eachScalar = extractTorchScalar(givenLoc, indexOfEachScalar,
+                                          given1DTensor, rewriter);
+    runningScalarSublist.push_back(eachScalar);
   }
-  return ValueList;
+
+  auto someTorchScalarType = getTorchScalarType(some1DTensorType, rewriter);
```
[the remark this answers is not in the document]
Comment (bjacobgordon) [reply lib/Conversion/TorchOnnxToTorch/DefaultDomainQtoZ.cpp]: Will do!

**Comment by bjacobgordon** [comment]
@zjgarvey All the tweaks were super quick, and one of them sparked a few ideas! It took some time, but I think it'll be worth it:
- Was able to extract out two more helpers that I think you'll like
- standardized the way I named their signatures so it'd be easier to modify for use with `ImplicitLocOpBuilder`
   - kept `rewriter` (and `givenLoc`) as the first param(s) each time 
   - kept the convention of rewriter methods starting with "get"/"create" for types/values (while still ensuring it was obvious what the method does)

I think there's a ton of opportunity for subsequent PRs to leverage these new helpers to gradually distill the implementation for these conversions!

Hoping that it leaves you nitless, but we'll see! Just let me know haha.

**Review by zjgarvey — CHANGES_REQUESTED** [review]
Hey, Jacob. 

The PR was in a great state to merge before you made the undiscussed changes. To make this process more efficient, would you please revert those changes so that we can merge the PR as discussed earlier?

This PR is blocking your progress on another PR, so I don't want to spend another day or two going back and forth on these new changes.

**Review by zjgarvey — APPROVED** [review]
Looks good to me. When <https://github.com/llvm/torch-mlir/pull/4008> lands, sync with main and we can merge.
